cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -26,321 +30,198 @@ where
+        |co| async move { yield_blocks(connector, &co, reference_timestamp).await }
+    });
 
     loop {
-        tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
-
-        match connector.latest_block().compat().await {
-            Ok(Some(block)) if block.hash.is_some() => {
-                let blockhash = block.hash.expect("cannot fail");
-
-                if !sent_blockhashes.contains(&blockhash) {
-                    sent_blockhashes.insert(blockhash);
-
-                    let _ = join(
-                        block_queue.send(block.clone()),
-                        find_parent_queue.send((blockhash, block.parent_hash)),
-                    )
-                    .await;
-
-                    if sent_blockhashes.len() == 1 {
-                        let _ = look_in_the_past_queue.send(block.parent_hash).await;
-                    };
+        match block_generator.async_resume().await {
+            GeneratorState::Yielded(block) => {
+                if let Some(transaction_and_receipt) =
+                    check_block_against_pattern(connector.clone(), block, pattern.clone()).await?
+                {
+                    return Ok(transaction_and_receipt);
+                } else {
+                    continue;
                 }
             }
-            Ok(Some(_)) => {
-                log::warn!("Ignoring block without blockhash");
-            }
-            Ok(None) => {
-                log::warn!("Could not get latest block");
-            }
-            Err(e) => {
-                log::warn!("Could not get latest block: {:?}", e);
-            }
-        };
+            GeneratorState::Complete(Err(e)) => return Err(e),
+            GeneratorState::Complete(Ok(infallible)) => match infallible {},
+        }
     }
 }
 
-/// Processes block hashes from the `next_hash` receiver, fetches blocks from
-/// the blockchain connector by block hash and enqueues the block on the
-/// `block_queue`. Enqueues the block hash and parent block hash on the
-/// `find_parent_queue`.
-async fn process_blocks_by_hash<C, E>(
-    connector: C,
-    mut block_queue: Sender<Block>,
-    mut find_parent_queue: Sender<(Hash, Hash)>,
-    next_block_channel: (Sender<Hash>, Receiver<Hash>),
-) where
+async fn yield_blocks<C, E>(
+    mut connector: C,
+    co: &Co<Block>,
+    reference_timestamp: Option<u32>,
+) -> anyhow::Result<std::convert::Infallible>
+where
     C: LatestBlock<Block = Option<Block>, Error = E>
         + BlockByHash<Block = Option<Block>, BlockHash = Hash, Error = E>
         + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash, Error = E>
         + Clone,
-    E: Debug + Send + 'static,
+    E: std::error::Error + Debug + Send + Sync + 'static,
 {
-    let (mut fetch_block_by_hash_queue, mut next_hash) = next_block_channel;
-    let mut sent_blockhashes: HashSet<H256> = HashSet::new();
+    let mut seen_blockhashes: HashSet<Hash> = HashSet::new();
```

Comment:
We currently have `blockhash` and `block_hash` in the code base.  Please consider adding a patch to the end of this PR converting the whole code base to use either `block_hash` or `blockhash`.